### ISSUE TITLE
If I am not mistaken find is linear, however we can do better

### ISSUE THREAD

**Issue by porkybrain** [edouarda/brigand#107]
If we transform<seq,pred> we essentially have a list of bools in O1, if we then create another list of bools with N true and the rest false (also O1), AND them (O1) and then compare the result to a list of all false tapes we can detect if there is a match in the half we anded with trues. 
This allows us to essentially do a "lower bounds" search rather than a linear search. If we use this pattern often and don't take size/2 but round up to the next power of 2 we would also achieve substantial re use of the list of bools and results of anding and comparing.

**Comment by porkybrain** [comment]
not sure if this will work in MSVC2013 but it does in MSVC2015
```C++
template<typename T, typename...Ts> //find index of T in Ts... in constant time
struct find_impl {
	using is = make_sequence<brigand::size_t<0>, sizeof...(Ts)>;
	using l = list<Ts...>;
	using zipped = transform<is, l, indexed_type<_1, _2>>;
	template<typename C>
	struct convert_me;
	template<template<typename...> class C, typename... Cs>
	struct convert_me<C<Cs...>> : Cs...{};
	template<typename I>
	static I convert(indexed_type<I, T>*) {};
	using type = decltype(convert(&std::declval<convert_me<zipped>>()));
};
```

**Comment by edouarda** [comment]
Interesting approach!  But how many interemediary types does it create? It also matters.

**Comment by ldionne** [comment]
How is this O(1)? You're instantiating n types (the `indexed_type`s), and we're counting the number of instantiations. Hence, this is O(n). That being said, it's much better than a recursive approach (at least on Clang), because deeply nested templates seems to be hard on the compiler.

**Comment by ldionne** [comment]
There are some patterns where this will be outperformed by the current implementation, though. See [this article](http://ldionne.com/2015/11/29/efficient-parameter-pack-indexing/) for benchmarks.

**Comment by edouarda** [comment]
What should we do then? Offer both? We can also use our "fast lane" approach to reduce recursion pressure.

**Comment by ldionne** [comment]
If I were you, I would benchmark the approach suggested by @porkybrain and the approach you currently use. I would then decide on one approach, and take the hit for the cases where the selected approach is worse. The user most likely doesn't care about choosing a particular implementation of `find` (and in general, he couldn't make a choice anyway).

**Comment by ldionne** [comment]
You might also want to benchmark a fast-laned version of your current recursive approach, if that's possible. I suspect that for `find`, however, it's impossible to have a fast-lane (because you want to stop whenever you find the right type).

**Comment by edouarda** [comment]
Yes but you can find by group of 10, dividing by 10 the cost. Or am I missing something?

**Comment by ldionne** [comment]
Perhaps it's possible, but I don't see how. How would you go about finding a group of 10 elements?

**Comment by edouarda** [comment]
I was thinking about something like

template <typename A, typename B, typename TheTypeIWant>
return_range operator()(A, B, TheTypeIWant) 
{
  position is 3
}

Wouldn't that work?

**Comment by porkybrain** [comment]
just to clarify, I'm just suggesting whatever I think of, I make no claim that it is well tested or better ;) If you were to peel off 10, perform a transform with the predicate and then compare the result to a list of 10 false_types (or just specialize for that) it should be faster than just plain recursion.

**Comment by edouarda** [comment]
Louis is correct when he says we should benchmark it.

**Comment by porkybrain** [comment]
Just to name a few more alternatives I have not seen before how about this (only works if each type in the list is unique):
```C++
template<bool B, std::size_t N>
struct TakeUpSpace
{
	int i[1];

};

template<std::size_t N>
struct TakeUpSpace<true, N>
{
	int i[N];
};

template<typename... T>
struct DeriveFrom : T...
{
};

template<typename T, typename U, typename V>
struct index_of_impl;

template<typename...T, std::size_t...I, typename V>
struct index_of_impl<list<T...>, std::index_sequence<I...>, V>
{
	using t = ;  //index_sequence starts at 0 and we already subtract one for each element
	using type = brigand::size_t<
		(
			sizeof(DeriveFrom<TakeUpSpace<std::is_same<V, T>::value, I + 2>...>) 
			/ sizeof(int)) 
		- sizeof...(T)>;
};

template<typename L, typename T>
using index_of = typename index_of_impl<L, std::make_index_sequence<size<L>::value>, T>::type;
```

**Comment by porkybrain** [comment]
this might be faster than you current "contains":
```C++
namespace match {
	struct yes : std::true_type {};
	template<typename...T>
	std::true_type test(T*...);
}

std::false_type test(...);

template<typename T, typename U>
struct same{
	using type = std::false_type*;
};

template<typename T>
struct same<T,T> {
	using type = match::yes*;
};

template<typename F, typename...T>
using contains = decltype(test(((typename same<T, F>::type)0)...));
```
If you combine mp_drop from [here](http://pdimov.com/cpp2/simple_cxx11_metaprogramming_2.html)  with a fast contains you could essentially implement a binary search (keep dropping different amounts untill you find where the result of contains changes)

**Comment by porkybrain** [comment]
Ok you might need this too or you could only find the last one (because you can only drop the front)
```C++
template<bool B, typename T, typename U>
struct same_and_use {
	using type = std::false_type*;
};

template<typename T>
struct same_and_use<true, T, T> {
	using type = match::yes*;
};

template<typename T, typename U, typename V, std::size_t N>
struct contained_before_index_impl;

template<typename T, typename... U, std::size_t...I, std::size_t N>
struct contained_before_index_impl<T,list<U...>,std::index_sequence<I...>,N> {
	using type = decltype(test(((typename same_and_use < (I<N), T, U>::type)0)...));
};


template<typename T, typename L, std::size_t N>  
using contained_before_index = typename contained_before_index_impl<T, L, std::make_index_sequence<size<L>::value>, N>::type;
```

**Comment by jfalcou** [comment]
>> Louis is correct when he says we should benchmark it.

we should have a set of benchmark for compile-time setup at some point.
@ldionne  any particular advice on what to use to do this in a sane way ?

**Comment by ldionne** [comment]
You can have a look at how Hana does it. I generate C++ files with Ruby templates, and then measure the compile-time it takes for a file to compile.

**Comment by porkybrain** [comment]
As a super non scientific but probably relevant note I have noticed when changing algorithms on 500 item plus lists is that it is often faster to look up types rather than generate them. This is only shown in tests if you compute several simmilar things (more like a real program) but not if you just test one thing. For example std::index_sequence of a certain length has a probability of being reused. 

Time vs. RAM use also seem to be two things, for example calculating something by exploiting ADL in a decltype context tends to use less RAM because the compiler can forget the template functions used but would keep them in ram if they were class templates.

**Comment by edouarda** [comment]
Louis when you said Ruby that part of me that thought you were awesome died.

**Comment by ldionne** [comment]
@porkybrain I'm not sure I understand what you mean. Can you please give a small example of what you mean by "it is often faster to look up types rather than generate them"?

@edouarda Too bad! Ruby has a very easy-to-use template engine named ERB. It's used by a lot of people to generate HTML pages, and I use it to generate C++ files. It's quite simple, but I think Python (and several other languages) also have similar facilities.

**Comment by edouarda** [comment]
@ldionne But you were my hero :-(

**Comment by edouarda** [comment]
I think you're right though we should use some scripting language to generate the test files. But documentation comes first.

**Comment by ldionne** [comment]
> I think you're right though we should use some scripting language to generate the test files. But documentation comes first.

Well, you can't really use the preprocessor or template metaprogramming to generate them, because that will influence the compile-time, which you're trying to measure. The only remaining option is to write the files by hand, which does not scale.

**Comment by edouarda** [comment]
Xzibit would tell you to write a C++ program that generates C++ programs to bench your C++ Metaprogramming library.

**Comment by edouarda** [comment]
Any update on this?

**Comment by porkybrain** [comment]
@ldionne in your article (about getting a type at a know index) you didn't benchmark recursive with a fast track: It probably shouldn't add much overhead to the normal fast track for small lists but it should be considerably faster for long lists.

Testing on clang I noticed that the current implementation is actually two recursions per element so clang dies at 128 elements with default settings. We certainly should be able to do better there.

**Comment by porkybrain** [comment]
```C++

	namespace detail
	{
		template<typename P, template <class...> class Sequence>
		struct finder {
			template<typename T>
			struct Pred : brigand::apply<P, T> {};

			//default only if not found
			template<bool C, bool F, typename...Ts>
			struct find {
				using type = Sequence<Ts...>;
			};

			//found it, terminate
			template<bool F, typename T1, typename...Ts>
			struct find<true, F, T1, Ts...> {
				using type = Sequence<T1, Ts...>;
			};

			//not found but close enough to no fast track
			template<typename T1, typename T2, typename...Ts>
			struct find<false, true, T1, T2, Ts...> : find<Pred<T2>::value,true,T2,Ts...> {};

			//not found and not enough to no fast track again
			template<typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9, typename...Ts>
			struct find<false, false, T1, T2, T3, T4, T5, T6, T7, T8, T9, Ts...> :
				find<Pred<T9>::value, true/*technically we don't know if its going to be in the next 8 but if not the default will kick in so its ok*/, T9, Ts...> {};

			//not found and can fast track again
			template<typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9, typename T10, typename T11, typename T12, typename T13, typename T14, typename T15, typename T16, typename...Ts>
			struct find<false, false, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, Ts...> :
				find<Pred<T9>::value, (Pred<T9>::value || Pred<T10>::value || Pred<T11>::value || Pred<T12>::value || Pred<T13>::value || Pred<T14>::value || Pred<T15>::value || Pred<T16>::value), T9, T10, T11, T12, T13, T14, T15, T16, Ts...> {};
		};
        // By default, consider we found nothing
        template <typename Predicate, class Sequence>
        struct find_if_impl
        {
            using type = Sequence;
        };

		template <typename Predicate, template <class...> class Sequence, typename T1, typename... T>
		struct find_if_impl<Predicate, Sequence<T1, T...>> : finder<Predicate, Sequence>::template find<false, false, void, void, void, void, void, void, void, void, T1, T... >	{};
    }
	namespace lazy
	{

		// find uses Predicate, it's very effective
		template <typename Sequence, typename Predicate = brigand::detail::non_null>
		using find = typename detail::find_if_impl<Predicate, Sequence>;
	}

	// find uses Predicate, it's very effective
	template <typename Sequence, typename Predicate = brigand::detail::non_null>
	using find = typename lazy::find<Sequence, Predicate>::type;

```

This should be as fast or up to 6x (or so) faster depending on the list length. More importantly with clang defaults you can find in a list which is 16x bigger than the current implementation. 

Doing 20 separate finds in a 1000 element list takes 2 seconds on my machine, probably fast enough for now. Current implementation takes 15s for 10 finds and crashes if I do 11 (not sure why, they are not dependent on each other).

**Comment by porkybrain** [comment]
MSVC2013 will not work with lists larger than 250 elements using the current implementation, 1000 is fine this one, doing 20 searches on a 1000 element list takes 16s on my machine.

Interestingly MSVC2015 hits a ceiling at about 350 element lists where efficiency just  goes down hill fast while MSVC2013 does not have this problem. Since the old algorithm couldn't get to 350 any way the new one is still better just not by as much as on other compilers.

As I mentioned above clang does much better with the new algorithm.

GCC results are essentially identical to clang.

In conclusion this seems better than what we have right now.

**Comment by porkybrain** [comment]
see #126
